src/katgpucbf/xbgpu/engine.py

```diff
@@ -909,19 +1085,21 @@ async def _receiver_loop(self) -> None:
         for pipeline in self._pipelines:
             pipeline.shutdown()
 
-    def _request_pipeline(self, stream_name: str) -> Pipeline:
+    def _request_pipeline(self, stream_name: str) -> tuple[Pipeline, int]:
         """Get the pipeline related to the katcp request.
 
-        Return the first Pipeline that matches by name.
+        Return the first Pipeline that matches by name, as well as the index
+        of the data-stream it transmits.
 
         Raises
         ------
         FailReply
             If the `stream_name` is not a known output.
         """
         for pipeline in self._pipelines:
-            if stream_name == pipeline.output.name:
-                return pipeline
+            for output in pipeline.outputs:
+                if stream_name == output.name:
+                    return pipeline, pipeline.outputs.index(output)
```

Comment:
This is searching the list of outputs twice. How about:
```suggestion
            for i, output in enumerate(pipeline.outputs):
                if stream_name == output.name:
                    return pipeline, i
```